extensions/HyperVExtension/src/HyperVExtension/Models/VirtualMachineCreation/VMGalleryVMCreationOperation.cs

```diff
@@ -179,34 +208,52 @@ private void UpdateProgress(string localizedString, uint percentage = 0u)
     /// <summary>
     /// Downloads the disk image from the Hyper-V VM gallery.
     /// </summary>
-    private async Task DownloadImageAsync()
+    private async Task DownloadImageAsync(VMGalleryImage image)
     {
         var downloadUri = new Uri(Image.Disk.Uri);
         var archivedFileName = _vmGalleryService.GetDownloadedArchiveFileName(Image);
         var archivedFileAbsolutePath = Path.Combine(_tempFolderSaveLocation, archivedFileName);
+        var isFileBeingDownloaded = _downloaderService.IsFileBeingDownloaded(archivedFileAbsolutePath);
 
         // If the file already exists and has the correct hash, we don't need to download it again.
-        if (File.Exists(archivedFileAbsolutePath))
+        if (File.Exists(archivedFileAbsolutePath) && !isFileBeingDownloaded)
         {
             ArchivedFile = await StorageFile.GetFileFromPathAsync(archivedFileAbsolutePath);
+            UpdateProgress(_stringResource.GetLocalized("VerifyingHashForExistingFile", $"({image.Name})"));
             if (await _vmGalleryService.ValidateFileSha256Hash(ArchivedFile))
             {
                 return;
             }
 
             // hash is not valid, so we'll delete/overwrite the file and download it again.
             _log.Information("File already exists but hash is not valid. Deleting file and downloading again.");
+            UpdateProgress(_stringResource.GetLocalized("HashVerificationFailed", $"({image.Name})"));
             await DeleteFileIfExists(ArchivedFile!);
         }
 
+        UpdateProgress(_stringResource.GetLocalized("DownloadStarting", $"({image.Name})"));
         await _downloaderService.StartDownloadAsync(this, downloadUri, archivedFileAbsolutePath, CancellationTokenSource.Token);
 
+        if (!HasDownloadCompleted())
+        {
+            // Wait for the last download to complete before moving on.
+            _waitForDownloadCompletion.WaitOne();
```

Comment:
We should think about adding a way (and UX) to cancel this download. It can get stuck forever.

Reply:
Yea I thought about adding a timeout, but I figure it depends on the users internet connection so I can't really say for sure how long it'll take. I'll make a GitHub issue for this to update the UI. We need to get the cancellation token from the `IAsyncOperation<CreateComputeSystemResult> StartAsync()` method above and use it here to cancel the operation.

Reply:
Good news, looks like I actually originally do send a cancelation request from Dev Home to the extension here:

https://github.com/microsoft/devhome/blob/7b65abfd6d373cd8d220f4182b5e044811e2631d/common/Environments/Models/CreateComputeSystemOperation.cs#L120

So I have updated the code to allow for cancellation: See below video as we now only use async/awaits:

https://github.com/user-attachments/assets/363e3ab7-0eff-4654-b23a-1eaeb8003388